src/main/scala/firrtl/analyses/InstanceKeyGraph.scala

```diff
@@ -52,8 +85,33 @@ class InstanceKeyGraph(c: ir.Circuit) {
     */
   def findInstancesInHierarchy(module: String): Seq[Seq[InstanceKey]] = {
     val instances = vertices.filter(_.module == module).toSeq
-    instances.flatMap{ i => fullHierarchy.getOrElse(i, Nil) }
+    instances.flatMap{ i => cachedFullHierarchy.getOrElse(i, Nil) }
   }
+
+  /** Given a circuit, returns a map from module name to a map
+    * in turn mapping instances names to corresponding module names
+    */
+  def getChildInstanceMap: mutable.LinkedHashMap[OfModule, mutable.LinkedHashMap[Instance, OfModule]] =
+    mutable.LinkedHashMap(childInstances.map { case (k, v) =>
+      val moduleMap: mutable.LinkedHashMap[Instance, OfModule] = mutable.LinkedHashMap(v.map(_.toTokens):_*)
+      TargetToken.OfModule(k) -> moduleMap
+    }:_*)
+
+  /** All modules in the circuit reachable from the top module */
+  def reachableModules: Seq[OfModule] = cachedReachableModules
+
+  /** All modules *not* reachable from the top module of the circuit */
+  def unreachableModules: Seq[OfModule] = cachedUnreachableModules
+
+  /** An [[firrtl.graph.EulerTour EulerTour]] representation of the [[firrtl.graph.DiGraph DiGraph]] */
+  private lazy val tour = EulerTour(internalGraph, circuitTopInstance)
+
+  /** Finds the lowest common ancestor instances for two module names in a design */
+  def lowestCommonAncestor(moduleA: Seq[InstanceKey], moduleB: Seq[InstanceKey]): Seq[InstanceKey] =
+    tour.rmq(moduleA, moduleB)
+
+  /** A list of absolute paths (each represented by a Seq of instances) of all module instances in the Circuit. */
+  def fullHierarchy: mutable.LinkedHashMap[InstanceKey, Seq[Seq[InstanceKey]]] = cachedFullHierarchy
```

Comment:
Similarly would rather use immutability here in 2.13

Reply:
Yes. But again, we still have to support `2.12` and `2.13` and I think it is beyond the scope of this PR to add special methods for `2.13`.
If you want the return to be immutable, I can return a `Seq[(K,V)]` instead.

Reply:
The introduction of a replacement API like `InstanceKeyGraph` is the only time we can fix mutable APIs. Specialization for 2.13 is annoying but gives the "best" API, since we could have `SeqMap` there and not have to lose the "map" aspect of it. Better than `Seq` would be `Iterable` because I think we could refine the API in 2.13.

Anyway, I won't insist here, but I think we should really consider not returning mutable types in 2.13 if we can.